src/tactic/ring2.lean

```diff
@@ -63,29 +79,42 @@ def eval {α} [comm_semiring α] (t : tree α) : csring_expr → α
 
 end csring_expr
 
+/-- An efficient representation of expressions in a commutative
+semiring using the sparse Horner normal form. This type admits
+non-optimal instantiations (e.g. `P` can be represented as `P+0+0`),
+so care must be taken to maintain an optimal, *canonical* form.-/
 @[derive decidable_eq]
 inductive horner_expr
+/- (const n) is a constant n in the csring, similarly to the same
+constructor in `csring_expr`. This one, however, can be negative. -/
 | const : znum → horner_expr
+/- (horner a x n b) is a*xⁿ + b, where x is the x-th atom
+in the atom tree. -/
 | horner : horner_expr → pos_num → num → horner_expr → horner_expr
 
 namespace horner_expr
 
+/-- "Is this `horner_expr` a valid `csring_expr`?" Basically all constants
+must be non-negative. -/
```

Comment:
Can this be made declarative? "Tests whether a `horner_expr` is a valid `csring_expr`, i.e. whether all constants are non-negative."